pkg/dynamicaccess/controller.go

```diff
@@ -0,0 +1,283 @@
+}
+
+// Limitation: If an upadate is called again within a second from the latest upload/update then mantaray save fails with ErrInvalidInput,
+// because the key (timestamp) is already present, hence a new fork is not created
+func (c *ControllerStruct) UpdateHandler(
+	ctx context.Context,
+	ls file.LoadSaver,
+	gls file.LoadSaver,
+	encryptedglRef swarm.Address,
+	historyRef swarm.Address,
+	publisher *ecdsa.PublicKey,
+	addList []*ecdsa.PublicKey,
+	removeList []*ecdsa.PublicKey,
+) (swarm.Address, swarm.Address, swarm.Address, swarm.Address, error) {
+	history, act, err := c.getHistoryAndAct(ctx, ls, historyRef, publisher, time.Now().Unix())
+	if err != nil {
+		return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+	}
+
+	gl, err := c.getGranteeList(ctx, gls, encryptedglRef, publisher)
+	if err != nil {
+		return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+	}
+	if len(addList) != 0 {
+		err = gl.Add(addList)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+	}
+	granteesToAdd := addList
+	if len(removeList) != 0 {
+		err = gl.Remove(removeList)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+		// generate new access key and new act, only if history was not newly created
+		if !historyRef.IsZero() {
+			act, err = c.newActWithPublisher(ctx, ls, publisher)
+			if err != nil {
+				return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+			}
+		}
+		granteesToAdd = gl.Get()
+	}
+
+	for _, grantee := range granteesToAdd {
+		err := c.accessLogic.AddGrantee(ctx, act, publisher, grantee)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+	}
+
+	granteeRef, err := gl.Save(ctx)
+	if err != nil {
+		return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+	}
+
+	egranteeRef, err := c.encryptRefForPublisher(publisher, granteeRef)
+	if err != nil {
+		return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+	}
+	// need to re-initialize history, because Lookup loads the forks causing the manifest save to skip the root node
+	if !historyRef.IsZero() {
+		history, err = NewHistoryReference(ls, historyRef)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+	}
+
+	mtdt := map[string]string{"encryptedglref": egranteeRef.String()}
+	hRef, actRef, err := c.saveHistoryAndAct(ctx, history, &mtdt, act)
+	if err != nil {
+		return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+	}
+
+	return granteeRef, egranteeRef, hRef, actRef, nil
+}
+
+func (c *ControllerStruct) Get(ctx context.Context, ls file.LoadSaver, publisher *ecdsa.PublicKey, encryptedglRef swarm.Address) ([]*ecdsa.PublicKey, error) {
```

Comment:
Every exported function needs a stardard format comment, please let this sink in

Reply:
Added comments to all exported functions and structs